google-cloud-tools-plugin/src/com/google/cloud/tools/intellij/appengine/sdk/CloudSdkServiceManager.java

```diff
@@ -43,6 +74,176 @@ public void onNewCloudSdkServiceTypeSelected(CloudSdkServiceType newServiceType)
     }
   }
 
+  /**
+   * Waits for Cloud SDK to be ready for all operations and then runs the given runnable. If process
+   * results in error or user cancel, shows notification and does not run.
+   *
+   * @param project Project to which runnable belongs.
+   * @param runnable Optional runnable to run after Cloud SDK is ready. This runnable will be run on
+   *     UI thread.
+   * @param progressMessage Message to show in progress dialog to identify which process is
+   *     starting, i.e. deployment or local server.
+   * @param callback Callback to report errors and log progress.
+   * @return blocking latch on which another thread may wait for Cloud SDK precondition checks to
+   *     finish.
+   */
+  public CountDownLatch runAfterCloudSdkPreconditionsMet(
```

Comment:
wdyt about getting rid of all of the "precondition" wording? Its really just doing a blocking install and then executing a callback after. How about:

- method: just `install` (I'm also second guessing this being here in the Manager instead of the service since the service has an `install` too. It may be better to just have this as an overloaded install method in the service itself. wdyt?)
- callback name: `SdkInstallCallback`. Callback variable name: `afterSdkInstall`.

Reply:
Well, `install` in service actually does installation of SDK if possible. This method could call `install` but only when a few factors are met, then it blocks waiting and executes code after that, again optionally. What it does all the time is runs task with progress, checks SDK status, shows errors and runs code if SDK is ready. maybe `runWhenSdkReady`?

Callback name is nice, thanks.

Also, I feel a bit bad ignoring return value in one place and using in another, what if CountDownLatch to call on completion would be passed in as an optional parameter instead?

Reply:
For some reason in my head I pictured you passing a runnable callback to the SdkService(/Manager) that would handle all the SDK business then execute the callback upon completion.

I see now that you are instead returning a latch that is used _by the caller_ to block until ready. (so my original comment here may not be applicable).

Lets take a closer look at this during the in person meeting